meadow-client/src/Semantics.purs

```diff
@@ -0,0 +1,1067 @@
+addToCommByTim timeout idCommit timData =
+  case M.lookup timeout timData of
+    Just commSet -> M.insert timeout (S.insert idCommit commSet) timData
+    Nothing -> M.insert timeout (S.singleton idCommit) timData
+
+-- Remove a commit identifier from the timeout data map 
+removeFromCommByTim :: Timeout -> IdCommit -> TimeoutData -> TimeoutData
+removeFromCommByTim timeout idCommit timData =
+  case M.lookup timeout timData of
+    Just commSet -> let newCommSet = S.delete idCommit commSet in
+                    if S.isEmpty newCommSet
+                    then M.delete timeout timData
+                    else M.insert timeout newCommSet timData
+    Nothing -> timData 
+
+-- Add a commit to CommitInfo
+addCommit :: IdCommit -> Person -> BigInt -> Timeout -> State -> State
+addCommit idCommit person value timeout (State state) =
+  State (state { commits = CommitInfo (ci { currentCommitsById =
+                                      M.insert idCommit
+                                               ({ person: person
+                                                , amount: value
+                                                , timeout: timeout })
+                                               ci.currentCommitsById
+                      , timeoutData = addToCommByTim timeout idCommit ci.timeoutData }) })
+  where (CommitInfo ci) = state.commits
+
+-- Return the person corresponding to a commit
+personForCurrentCommit :: IdCommit -> State -> Maybe Person
+personForCurrentCommit idCommit (State state) =
+  case M.lookup idCommit ci.currentCommitsById of
+    Just v -> Just v.person
+    Nothing -> Nothing
+  where (CommitInfo ci) = state.commits
+
+-- Check whether the commit is current (committed not expired)
+isCurrentCommit :: IdCommit -> State -> Boolean
+isCurrentCommit idCommit (State state) =
+  case M.lookup idCommit ci.currentCommitsById of
+    Just _ -> true 
+    Nothing -> false
+  where CommitInfo ci = state.commits
+
+-- Check whether the commit is expired
+isExpiredCommit :: IdCommit -> State -> Boolean
+isExpiredCommit idCommit (State state) = idCommit `S.member` ci.expiredCommitIds
+  where CommitInfo ci = state.commits
+
+-- Remove a current commit from CommitInfo
+removeCurrentCommit :: IdCommit -> State -> State
+removeCurrentCommit idCommit (State state) =
+  case M.lookup idCommit commById of
+    Just v -> State (state { commits = CommitInfo (ci { currentCommitsById = M.delete idCommit commById
+                                                      , timeoutData = removeFromCommByTim v.timeout idCommit timData }) })
+    Nothing -> State state 
+  where CommitInfo ci = state.commits
+        commById = ci.currentCommitsById
+        timData = ci.timeoutData
+
+-- Get expired not collected for person
+getRedeemedForPersonCI :: Person -> CommitInfo -> BigInt
+getRedeemedForPersonCI person (CommitInfo ci) =
+  case M.lookup person ci.redeemedPerPerson of
+    Nothing -> fromInt 0
+    Just value -> value
```

Comment:
`fromMaybe (fromInt 0) $ M.lookup person ci.redeemedPerPerson`?

Reply:
Indeed, "lookupDefault" would be ideal. Will fix.